cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java

```diff
@@ -207,10 +229,15 @@ public BasicMapReduceTaskContext load(ContextCacheKey key) throws Exception {
     private final Configuration configuration;
 
     private ContextCacheKey(TaskAttemptContext context) {
-      this.taskAttemptID = context.getTaskAttemptID();
-      this.configuration = context.getConfiguration();
+      this(context.getTaskAttemptID(), context.getConfiguration());
+    }
+
+    private ContextCacheKey(@Nullable TaskAttemptID taskAttemptID, Configuration configuration) {
```

Comment:
Regarding metrics collection, let's make sure it's documented. What would happen if the user attempts to emit metrics anyway?

It might not be a problem that the context would be shared. I was just wondering whether there can be anything troublesome. For example, if they getDataset(), they would share that instance of the dataset, right?